File: app/src/main/java/com/extendvr/TrackingImage.java
/*
* This class is an image/container class for easy retrevial and manipulation of the YCbCr image data
* */

package com.extendvr;

import android.support.annotation.ColorInt;
import android.util.Log;

public class TrackingImage {
    public byte[] Y;
    public byte[] Cb;
    public byte[] Cr;
    public int width = 320;
    public int height = 240;
    /* YCbCr images often have less color data compared to brightness data, to
    * take advantage of the fact that our eyes are more sensitive to brightness than
    * color. we store this in the trackingImage so that when we get the pixel data, we
    * can account for this.*/
    public int ColorPixelStride = 2;
    public int ColorRowStride = 2;

    public TrackingImage(byte[] y, byte[] cb, byte[] cr){
        Y = y;
        Cb = cb;
        Cr = cr;
    }

    @ColorInt
    public int getPixel(int x, int y){ // cb and cr have half the width and x/2 and y/2 because their pixel stride is 2, ie 1 pixel for every 4 Y pixels
        if(x < width && y < height)
        return Pixel.YCbCr(Y[x+y*320] & 0xff,
             Cb[x/ColorPixelStride+(y/ColorRowStride)*320/ColorPixelStride]& 0xff,
             Cr[x/ColorPixelStride+(y/ColorRowStride)*320/ColorPixelStride]& 0xff);
        else Log.e("ERROR","x or y outiside bounds: x is " + x + "y is " + y);
        return 0;
    }

}

File: app/src/main/java/com/extendvr/TrackingImage.kt
/*
* This class is an image/container class for easy retrevial and manipulation of the YCbCr image data
* */
package com.extendvr

import android.support.annotation.ColorInt
import android.util.Log
import com.extendvr.Pixel.YCbCr

class TrackingImage(var Y: UByteArray, var Cb: UByteArray, var Cr: UByteArray) {
    var width = 320
    var height = 240

    /* YCbCr images often have less color data compared to brightness data. I.E 1 color pixel for every 4 brightness pixels.
    * hence we need to store the color pixel stride and row stride, because there is gonna be a quarter the amount of data for color
    * */
    var ColorPixelStride = 2
    var ColorRowStride = 2

    @ColorInt
    fun getPixel(x: Int, y: Int): Int { // cb and cr have half the width and x/2 and y/2 because their pixel stride is 2, ie 1 pixel for every 4 Y pixels
        if (x < width && y < height && x >= 0 && y >= 0) return YCbCr(Y[x + y * 320].toInt(),
                Cb[x / ColorPixelStride + y / ColorRowStride * 320 / ColorPixelStride].toInt(),
                Cr[x / ColorPixelStride + y / ColorRowStride * 320 / ColorPixelStride].toInt()) else Log.e("ERROR", "x or y outiside bounds: x is " + x + "y is " + y)
        return -1 // pixel does not exist
    }

}